### PR TITLE
[SPARK-24613][SQL] Cache with UDF could not be matched with subsequent dependent caches

### PR DESCRIPTION
**Pull request by maryannxue**
## What changes were proposed in this pull request?

Wrap the logical plan with a `AnalysisBarrier` for execution plan compilation in CacheManager, in order to avoid the plan being analyzed again.

## How was this patch tested?

Add one test in `DatasetCacheSuite`

### CODE REVIEW

**Review by gatorsmile — COMMENTED** [review]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetCacheSuite.scala [hunk 1]
```diff
@@ -132,4 +132,19 @@ class DatasetCacheSuite extends QueryTest with SharedSQLContext with TimeLimits
     df.unpersist()
     assert(df.storageLevel == StorageLevel.NONE)
   }
+
+  test("SPARK-24613 Cache with UDF could not be matched with subsequent dependent caches") {
+    val expensiveUDF = udf({x: Int => Thread.sleep(10000); x})
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetCacheSuite.scala]: can we use accumulators and make sure this UDF only run 10 times? sleeping 10 seconds is not good in a unit test

Comment (maryannxue) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetCacheSuite.scala]: Accumulators probably wouldn't work. I'll do verify plan though.

**Comment by gatorsmile** [comment]
Thanks! Merged to master/2.3

**Comment by gatorsmile** [comment]
This is also a regression. Backported to 2.3 branch too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala [hunk 2]
```diff
@@ -142,7 +142,7 @@ class CacheManager extends Logging {
         // Remove the cache entry before we create a new one, so that we can have a different
         // physical plan.
         it.remove()
-        val plan = spark.sessionState.executePlan(cd.plan).executedPlan
+        val plan = spark.sessionState.executePlan(AnalysisBarrier(cd.plan)).executedPlan
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: We need a test case to cover this change.

**Pull request closed without merging**